pkg/orchestrator/orchestrator.go

```diff
@@ -0,0 +1,169 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package orchestrator
+
+import (
+	"context"
+	"time"
+
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+
+	"go.opentelemetry.io/auto/pkg/errors"
+	"go.opentelemetry.io/auto/pkg/instrumentors"
+	"go.opentelemetry.io/auto/pkg/log"
+	"go.opentelemetry.io/auto/pkg/opentelemetry"
+	"go.opentelemetry.io/auto/pkg/process"
+)
+
+// Interface defines orchestrator interface.
+type Interface interface {
+	Run() error
+}
+
+type pidServiceName struct {
+	serviceName string
+	pid         int
+}
+
+type impl struct {
+	ctx         context.Context
+	analyzer    *process.Analyzer
+	targetArgs  *process.TargetArgs
+	processch   chan *pidServiceName
+	deadProcess chan int
+	managers    map[int]*instrumentors.Manager
+	exporter    sdktrace.SpanExporter
+	pidTicker   <-chan time.Time
+}
+
+// New creates a new Implementation of orchestrator Interface.
+func New(
+	ctx context.Context,
+	targetArgs *process.TargetArgs,
+	exporter sdktrace.SpanExporter,
+) (Interface, error) {
+	return &impl{
+		ctx:         ctx,
+		analyzer:    process.NewAnalyzer(),
+		targetArgs:  targetArgs,
+		exporter:    exporter,
+		processch:   make(chan *pidServiceName, 10),
+		deadProcess: make(chan int, 10),
+		managers:    make(map[int]*instrumentors.Manager),
+		pidTicker:   time.NewTicker(2 * time.Second).C,
+	}, nil
+}
+
+func (i *impl) Run() error {
+	go i.findProcess()
+	for {
+		select {
+		case <-i.ctx.Done():
+			log.Logger.Info("Got SIGTERM cleaning up")
+
+			for _, m := range i.managers {
+				m.Close()
+			}
+
+			close(i.deadProcess)
+			close(i.processch)
+
+			return nil
+		case d := <-i.deadProcess:
+			log.Logger.Info("process died cleaning up", "pid", d)
```

Comment:
Need to clean BPFFS for dead processes. This can be done by calling `bpffs.Cleanup` with the dead PID